### PR TITLE
Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76477
* #76360

Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

### CODE REVIEW

File: test/distributed/_shard/sharded_tensor/ops/test_softmax.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    def _test_sharded_softmax(self, dim):
+        torch.manual_seed(0)
+        local_tensor = torch.rand(10, 10, device=self.rank)
+        local_softmax = torch.nn.functional.softmax(local_tensor, dim)
+
+        spec = ChunkShardingSpec(dim=0, placements=[f'rank:{idx}/cuda:{idx}' for idx in range(self.world_size)])
+        st = _shard_tensor(local_tensor, spec)
+        sharded_softmax = torch.nn.functional.softmax(st, dim)
+
+        self.assertEqual(local_softmax.chunk(self.world_size)[self.rank], sharded_softmax.local_tensor())
```
Comment (fduwjj) [inline test/distributed/_shard/sharded_tensor/ops/test_softmax.py:40]: Nit: you can just use _shard_tensor and directly compare two sharded tensor.

File: torch/distributed/_shard/sharded_tensor/partial_tensor.py [hunk 2]
```diff
@@ -152,3 +184,61 @@ def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> ShardedTensor:
...
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        if func in _PARTIAL_TENSOR_OPS:
+            return _PARTIAL_TENSOR_OPS[func](types, args, kwargs)
+
+        raise RuntimeError(
+            f"torch function '{func.__name__}', with args: {args} and "
+            f"kwargs: {kwargs} not supported for PartialTensor!")
+
+    def transpose(self, dim0, dim1):
```
Comment (fduwjj) [inline torch/distributed/_shard/sharded_tensor/partial_tensor.py:200]: Maybe after you land this change, I can make Partial Tensor a subclass of tensor too.

File: torch/distributed/_shard/replicated_tensor.py [hunk 3]
```diff
@@ -5,6 +5,14 @@
 from torch.distributed import distributed_c10d
 from torch.overrides import get_default_nowrap_functions
 
+_REPLICATED_WITH_NON_TENSOR_ALLOWLIST = [
+    # List of ops where if parameters are a combination of ReplicatedTensors
+    # and non-tensors, we can still return a ReplicatedTensor as the result.
+    torch.unsqueeze,
+    torch.Tensor.unsqueeze,
+    torch.Tensor.__getitem__,
```
Comment (fduwjj) [inline torch/distributed/_shard/replicated_tensor.py:13]: This is for `replicated_tensor[0]`, right?

File: torch/distributed/_shard/replicated_tensor.py [hunk 4]
```diff
@@ -101,7 +112,8 @@ def dispatch_arg(arg):
             rs = func(*args, **kwargs)
             if func in get_default_nowrap_functions():
                 return rs
-            if all_replicated and isinstance(rs, torch.Tensor) and not isinstance(rs, cls):
+            if (all_replicated and isinstance(rs, torch.Tensor) and not isinstance(rs, cls)) or \
```
Comment (fduwjj) [inline torch/distributed/_shard/replicated_tensor.py:115]: N00b question, is the 'cls' here just `ReplicatedTensor`?

**Review by fduwjj — APPROVED** [review]
Overall, looks great and you might need to rebase the PR a little bit. Thanks!

File: torch/distributed/_shard/replicated_tensor.py [hunk 5]
```diff
@@ -5,6 +5,14 @@
 from torch.distributed import distributed_c10d
 from torch.overrides import get_default_nowrap_functions
 
+_REPLICATED_WITH_NON_TENSOR_ALLOWLIST = [
```
Comment (wanchaol) [inline torch/distributed/_shard/replicated_tensor.py:8]: I think this list is for the case where all args are either ReplicatedTensor or scalars, instead of having two flags `all_replicated` and `replicated_with_non_tensor`, could we have a single flag to check if there's no `torch.Tensor` in the args (ShardedTensor or PartialTensor handles in their own __torch_function__), then we convert to ReplicatedTensor, otherwise leave it as tensor?

**Review by wanchaol — APPROVED** [review]
looks good to me, one suggestion to simplify the return type conversion for ReplicatedTensor

File: test/distributed/_shard/sharded_tensor/ops/test_softmax.py (hunk 1 above)
Comment (pritamdamania87) [reply test/distributed/_shard/sharded_tensor/ops/test_softmax.py:40]: `.chunk` seems simpler for a test rather than sharding the local tensor :)

File: torch/distributed/_shard/replicated_tensor.py (hunk 5 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/replicated_tensor.py:8]: Hmm but couldn't there be a case where there is an op with ReplicatedTensor and PartialTensor and it gets dispatched to ReplicatedTensor's `torch_function` instead? We re-dispatch to ShardedTensor in the case of ShardedTensor, but I don't see anything similar for ReplicatedTensor.

File: torch/distributed/_shard/replicated_tensor.py (hunk 3 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/replicated_tensor.py:13]: Yes more generally something like `replicated_tensor[...]`

File: torch/distributed/_shard/replicated_tensor.py (hunk 4 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/replicated_tensor.py:115]: Yes, let me update it so it is more clear.

File: torch/distributed/_shard/sharded_tensor/partial_tensor.py [hunk 6]
```diff
@@ -1,13 +1,45 @@
...
+    def decorator_sharded_func(wrapped_func):
+        _register_partial_tensor_op(func, wrapped_func)
+
+        @functools.wraps(wrapped_func)
+        def wrapper(*args, **kwargs):
+            return wrapped_func(*args, **kwargs)
+        return wrapper
+    return decorator_sharded_func
 
 class _PartialTensor(object):
```
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/_shard/sharded_tensor/partial_tensor.py:44]: Yup will rebase this.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    input = args[0]
+    pg = input._process_group
+    dim = kwargs['dim']
+    sharding_dim = input.sharding_spec().dim
+    ndims = len(input.size())
+    if dim == sharding_dim or dim + ndims == sharding_dim or sharding_dim + ndims == dim:
+        exp = torch.exp(input.local_tensor())
+        exp_sum = exp.sum(dim=dim).unsqueeze(dim=dim)
+        exp_sum = torch.distributed.nn.functional.all_reduce(exp_sum, group=pg)
+        smax = torch.div(exp, exp_sum)
```
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py:19]: Yup will do as I rebase.

File: torch/distributed/_shard/replicated_tensor.py (hunk 5 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/replicated_tensor.py:8]: So another issue is Replicated (op) scalar is not always guaranteed to be ReplicatedTensor. Imagine torch.add(ReplicatedTensor, x) where `x` is different on each rank.

So it seems like we do need this additional flag in which case we have look through a whitelist like this.
